src/lib/input/input-container.ts

```diff
@@ -232,24 +226,51 @@ export class MdInputDirective {
   constructor(private _elementRef: ElementRef,
               private _renderer: Renderer2,
               private _platform: Platform,
+              private _changeDetectorRef: ChangeDetectorRef,
               @Optional() @Self() public _ngControl: NgControl,
               @Optional() private _parentForm: NgForm,
               @Optional() private _parentFormGroup: FormGroupDirective,
               @Optional() @Inject(MD_ERROR_GLOBAL_OPTIONS) errorOptions: ErrorOptions) {
 
     // Force setter to be called in case id was not specified.
     this.id = this.id;
-
     this._errorOptions = errorOptions ? errorOptions : {};
     this.errorStateMatcher = this._errorOptions.errorStateMatcher || defaultErrorStateMatcher;
   }
 
-  /** Focuses the input element. */
-  focus() { this._elementRef.nativeElement.focus(); }
+  ngOnChanges() {
+    this._stateChanges.next();
+  }
 
-  _onFocus() { this.focused = true; }
+  ngOnDestroy() {
+    this._stateChanges.complete();
+  }
 
-  _onBlur() { this.focused = false; }
+  ngDoCheck() {
+    if (this._ngControl) {
+      // We need to re-evaluate this on every change detection cycle, because there are some
+      // error triggers that we can't subscribe to (e.g. parent form submissions). This means
+      // that whatever logic is in here has to be super lean or we risk destroying the performance.
```

Comment:
Would it make sense to also add a note to the Custom Error Matcher section of input.md that mentions this?

*Note: An input's error state must be calculated every change detection cycle, so it is important to keep the logic in your error state matcher very lean or you risk destroying performance*.